### PR TITLE
Fully Working Prototype with Bazel

### PR DESCRIPTION
**Pull request by lighterm**

### CODE REVIEW

**Comment by pratyai** [comment]
Thanks! Could you please format the .java files with https://github.com/google/google-java-format?

File: README.md [hunk 1]
```diff
@@ -32,3 +28,57 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+
+## Overview
+
+Java Function Coverage is a java agent that instruments the specified classes before load time. It collects the function coverage data for all the methods in those classes. The agent must be passed to the jvm with -javaagent option. The agent will instrument the bytecodes of the classes before they load into memory. 
+
+## Quickstart
+
+```bash
+# Get the package from Github and install it from main directory of the package
+$ mvn package
```
Comment (pratyai) [inline README.md]: We could drop mvn (configs and instructions), I think.

File: README.md [hunk 2]
```diff
@@ -32,3 +28,57 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+    * If its jar path/to/handler must be a path to a jar file  
+    * If its dir path/to/handler must be a path to a directory that contains .class files (folders must be structured same as jar files)
+    
+* packageName.ClassName argument must be the fully qualified name of the Handler class (example.handler.Handler in the sample Handler.java) that contains a start() method and necessary constructor. An instance of this class will be created during runtime and start() method will be called at the very beginning. See the Handler directory for example implementations.  
+
+```bash
+$  java -javaagent:path/to/agent.jar="type:path/to/handler packageName.ClassName" [other args..]
+```
+
+### Example Usage with Bazel
```
Comment (pratyai) [inline README.md]: Building with Bazel should be the canonical way (instead of mvn). I think the Quickstart section could go something like:

## Quickstart

### With the included examples
* Build the funccover agent < link to concepts > :
< bulid command >
* Build the example handler < link to concepts >:
< build command >
* Build the example program:
< build command >
* Run the program:
< run command >

### With customization
* Build the funccover agent < link to concepts > :
< bulid command >
* Write an handler < link to concepts > :
< briefly explain how to write one, link the example handler >
* Build the program:
< just explain that the user can use whatever prebuild binary they have >
* Run the program:
< run command >

### Concepts
< overview, explain what's an agent,  what's a handler, etc. >

**Review by pratyai — CHANGES_REQUESTED** [review]

**Review by cher11 — APPROVED** [review]

File: src/main/java/com/funccover/CoverageTransformer.java [hunk 3]
```diff
@@ -1,121 +1,125 @@
...
+  // the ClassPool object may not be able to find user classes
+  // In that case, an additional class path must be registered to the ClassPool.
+  // ClassPool used to compile inserted source code to bytecode
+  private final ClassPool classPool = ClassPool.getDefault();
+
+  CoverageTransformer() {
+    classPool.importPackage("com.funccover.CoverageMetrics");
+  }
+
+  // transform instruments given bytecode and returns instrumented bytecode
```
Comment (pratyai) [inline src/main/java/com/funccover/CoverageTransformer.java]: s/transform/Transform/ (i.e. capitalize); end the line with a period (i.e. ".").

File: src/main/java/com/funccover/CoverageTransformer.java [hunk 4]
```diff
@@ -1,121 +1,125 @@
...
+  // if it returns null, then given class will be loaded without instrumentation
```
Comment (pratyai) [inline src/main/java/com/funccover/CoverageTransformer.java]: s/if/If/ (i.e. capitalize); end the line with a period (i.e. ".").

File: src/main/java/com/funccover/CoverageTransformer.java [hunk 5]
```diff
@@ -1,121 +1,125 @@
...
+  @Override
+  public byte[] transform(
+      ClassLoader loader,
+      String className,
+      Class<?> classBeingRedefined,
+      ProtectionDomain protectionDomain,
+      byte[] classfileBuffer)
+      throws IllegalClassFormatException {
+
+    // if we do not want to instrument given class, return null
```
Comment (pratyai) [inline src/main/java/com/funccover/CoverageTransformer.java]: s/if/If/ (i.e. capitalize); end the line with a period (i.e. ".").

File: src/main/java/com/funccover/CoverageTransformer.java [hunk 6]
```diff
@@ -1,121 +1,125 @@
...
-
 import javassist.CannotCompileException;
 import javassist.ClassPool;
-import javassist.CtBehavior;
 import javassist.CtClass;
 import javassist.CtMethod;
 import javassist.Modifier;
 
-
 // CoverageTransformer implements a class that will be used in instrumentation
```
Comment (pratyai) [inline src/main/java/com/funccover/CoverageTransformer.java]: Please capitalize the first letter of the sentences in the comment and end with a period throughout all the files involved.

File: src/main/java/com/funccover/CoverageTransformer.java [hunk 7]
```diff
@@ -1,121 +1,125 @@
...
+        flag = true;
+      }
+
+      if (flag) {
+        result = ct.toBytecode();
+      }
+
+      // detach removes newly created class from cp to avoid unnecesarry memory consumption
+      ct.detach();
+    } catch (Throwable e) {
```
Comment (pratyai) [inline src/main/java/com/funccover/CoverageTransformer.java:100]: * Why "Throwable" instead of a more specific exception type?
* Why catch the exception at all? Are there "noise" exceptions where the behaviour of the instrumented program is not impacted?

File: src/main/java/com/funccover/HandlerLoader.java [hunk 8]
```diff
@@ -1,120 +1,119 @@
...
-        }
-        System.out.println("invalid path type " + pathArgs[0]);
-        return null;
+    if (handler == null || start == null || cl == null) {
+      return;
     }
 
+    try {
+      start.invoke(handler);
+    } catch (Exception e) {
```
Comment (pratyai) [inline src/main/java/com/funccover/HandlerLoader.java:74]: Same comment about catching exceptions as the other comment.

**Review by pratyai — CHANGES_REQUESTED** [review]

File: src/main/java/com/funccover/CoverageTransformer.java (hunk 6 above)
Comment (lighterm) [reply src/main/java/com/funccover/CoverageTransformer.java]: Last comment fix commit resolves this issue.

File: src/main/java/com/funccover/CoverageTransformer.java (hunk 3 above)
Comment (lighterm) [reply src/main/java/com/funccover/CoverageTransformer.java]: here the name of the method is transform, should I make it Transform?

Comment (pratyai) [reply src/main/java/com/funccover/CoverageTransformer.java]: Right; didn't notice it's the function name. Put in quotes, like `transform`, I guess.

File: WORKSPACE [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+maven_install(
+    artifacts = [
+        "org.javassist:javassist:3.23.1-GA",
+    ],
+    repositories = [
+        "https://jcenter.bintray.com/",
+        "https://repo1.maven.org/maven2",
+    ],
+)
```
Comment (pratyai) [inline WORKSPACE]: Could you run buildifier (https://github.com/bazelbuild/buildtools) over bazel related files?

**Review by pratyai — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (lighterm) [reply README.md]: Can you check it, I updated readme.md

File: WORKSPACE (hunk 9 above)
Comment (lighterm) [reply WORKSPACE]: fixed

File: src/main/java/com/funccover/CoverageTransformer.java (hunk 7 above)
Comment (lighterm) [reply src/main/java/com/funccover/CoverageTransformer.java:100]: Well, I thought it would be better not exit instrumentation if it fails somewhere. I can change it. I don't think there are noise exceptions that don't mean anything.

File: README.md [hunk 10]
```diff
@@ -32,3 +28,86 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+```bash
+$ bazel build //src/main/java/com/funccover:funccover_deploy.jar
+```
+
+* Build the example [handler](#handler) in the repository
+
+```bash
+$ bazel build //src/main/java/example/handler:Handler
+```
+
```
Comment (pratyai) [inline README.md:50]: Add a line here:
This should generate the file `bazel-bin/src/main/java/example/handler/libHandler.jar` under the project root.

File: README.md [hunk 11]
```diff
@@ -32,3 +28,86 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+
+### With The Included Examples
+
+* Build [funccover agent](#funccover-agent)
+
+```bash
+$ bazel build //src/main/java/com/funccover:funccover_deploy.jar
+```
+
```
Comment (pratyai) [inline README.md:42]: Add a line here:
This should generate the file `bazel-bin/src/main/java/com/funccover/funccover_deploy.jar` under the project root.

File: README.md [hunk 12]
```diff
@@ -32,3 +28,86 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+* Build the example [handler](#handler) in the repository
+
+```bash
+$ bazel build //src/main/java/example/handler:Handler
+```
+
+* Build the example program binary with coverage
+
+```bash
+$ bazel build --jvmopt="-javaagent:agent_path='jar:handler_path example.handler.Handler' "  //src/main/java/example/program:HelloWorld
```
Comment (pratyai) [inline README.md]: I think agent_path and handler_path can be replaced with "bazel-bin/src/main/..." versions to allow just copying it (well, that works under the assumption that the command is run from the root; but that would be obvious to the reader at this point, I think).

File: src/main/java/com/funccover/CoverageTransformer.java (hunk 7 above)
Comment (pratyai) [reply src/main/java/com/funccover/CoverageTransformer.java:100]: I'm undecided about whether to abort to let the user know that the instrumentation didn't work, or to just continue. Let's keep it as it (i.e. with catching & ignoring exceptions) is for now.

**Review by pratyai — APPROVED** [review]

**Comment by pratyai** [comment]
Please resolve the remaining comments; and let me know so that I can merge it. Thanks!

File: README.md (hunk 12 above)
Comment (lighterm) [reply README.md]: done

File: README.md (hunk 11 above)
Comment (lighterm) [reply README.md:42]: done

File: README.md (hunk 10 above)
Comment (lighterm) [reply README.md:50]: done

**Comment by pratyai** [comment]
@muratekici One last request; could you please reformat the java files? The format checker is complaining. Thanks!

**Comment by lighterm** [comment]
Checkers are fine now I think.
